## Python Lists - Part 2

Python knows a number of <i>compound</i> data types, used to group together other values. The most versatile is the <b>list</b>, which can be written as a list of comma-separated values (items) between square brackets. Lists might contain items of different types, but usually the items all have the same type.

A <b>list</b> stores a series of items in a particular order. You access items using an index, or within a loop.

Lists allow you to store sets of information in one place, whether you have just a few items or millions of items. Lists are one of Python's most powerful features readily accessible to new programmers, and they tie together many important concepts in programming.

Lists are a <b>Mutable</b> type, it is possible to change their content.

### Creating a List / Defining a List / Making a List

Use square brackets to create a list, and use commas to separate individual items in the list.

In [1]:
users = []

In [2]:
values = [5, 2, 4, 8, 7]

In [3]:
prog_langs = [['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]] 

In [4]:
data = list([])

### Adding / Appending Elements to a List

You can add elements to the end of a list.

<code>append(value)</code> – appends a new element to the end of the list.

In [5]:
users.append?

Signature: users.append(object, /)
Docstring: Append object to the end of the list.
Type:      builtin_function_or_method


In [6]:
users.append('Ahmad')
users.append('Bilal')
users.append('Khalil')

In [7]:
values.append(4)

In [8]:
prog_langs.append(['Java', 6])

### Accessing / Getting Elements from the List

Individual elements in a list are accessed according to their position, called the <i>index</i>. The index of the first element is 0, the index of the second element is 1, and so forth. Negative indices refer to items at the end of the list. To get a particular element, write the name of the list and then the index of the element in square brackets

Indexing allows you to access/change/delete only a single cell of a list.

Indexing 
   - Positive or Forward or Zero-based Indexing 
   - Negative or Backward or Reverse Indexing  

<b>Accessing List Elements by Positive Index Number</b>

In [9]:
print(users[0])   # indexing returns the item
print(users[1])
print(users[2])

Ahmad
Bilal
Khalil


In [10]:
# print(users[3])   # IndexError: list index out of range

In [11]:
print(values[0])
print(values[1])
print(values[2])
print(values[3])
print(values[4])
print(values[5])

5
2
4
8
7
4


In [12]:
print(values[len(values)-1])  # values[5]

4


In [13]:
print(values[4])

7


In [14]:
print(values[values[2]])

7


In [15]:
print(prog_langs[0])
print(prog_langs[1])
print(prog_langs[2])
print(prog_langs[3])
print(prog_langs[4])
print(prog_langs[5])

['Rust', 3]
['Python', 1]
['JavaScript', 2]
['Julia', 5]
['Go', 4]
['Java', 6]


In [16]:
print(prog_langs[0][0])
print(prog_langs[0][1])
print(prog_langs[1][0])
print(prog_langs[1][1])

Rust
3
Python
1


<b>Accessing List Elements by Negative Index Number</b>

In [17]:
print(users[-1])
print(users[-2])
print(users[-3])

Khalil
Bilal
Ahmad


In [18]:
# print(users[-4])   # IndexError: list index out of range

In [19]:
print(values[-6])

5


In [20]:
print(prog_langs[-3])
print(prog_langs[-3][0])
print(prog_langs[-3][1])

['Julia', 5]
Julia
5


### Slicing a List

You can work with any set of elements from a list. A portion of a list is called a <i>slice</i>.

Slice operation is performed on Lists with the use of colon(:). 

Slicing is a flexible tool to build new lists out of an existing list.

The full slice syntax is: <code>start:stop:step</code>. 

- <i>start</i> refers to the index of the element which is used as a start of our slice. 
- <i>stop</i> refers to the index of the element we should stop just before to finish our slice. 
- <i>step</i> allows you to take each nth-element within a <code>start:stop</code> range.

All slice operations return a new list containing the requested elements. 

In [21]:
print(users[::])

['Ahmad', 'Bilal', 'Khalil']


<b>List <code>Strat</code> Slicing</b>

In [22]:
print(users[0:])

['Ahmad', 'Bilal', 'Khalil']


In [23]:
print(users[1:])      # sublist of the list   ... # slicing returns a new list

['Bilal', 'Khalil']


In [24]:
print(users[2:])

['Khalil']


In [25]:
print(users[3:])

[]


In [26]:
print(users[-1:])

['Khalil']


In [27]:
print(users[-2:])

['Bilal', 'Khalil']


In [28]:
print(users[-3:])

['Ahmad', 'Bilal', 'Khalil']


In [29]:
print(users[-4:])  

['Ahmad', 'Bilal', 'Khalil']


<b>List <code>Start-Stop</code> Slicing</b>

In [30]:
print(users[:3])

['Ahmad', 'Bilal', 'Khalil']


In [31]:
print(users[:2])

['Ahmad', 'Bilal']


In [32]:
print(users[:1])

['Ahmad']


In [33]:
print(users[:0])

[]


In [34]:
print(users[0:])

['Ahmad', 'Bilal', 'Khalil']


In [35]:
print(users[:-1])

['Ahmad', 'Bilal']


In [36]:
print(users[:-2])

['Ahmad']


In [37]:
print(users[:-3])

[]


In [38]:
print(users[0:len(users)])

['Ahmad', 'Bilal', 'Khalil']


In [39]:
print(users[1:len(users)])

['Bilal', 'Khalil']


In [40]:
print(users[2:len(users)])

['Khalil']


In [41]:
print(users[3:len(users)])

[]


In [42]:
print(users[-1:len(users)])

['Khalil']


In [43]:
print(users[-2:len(users)])

['Bilal', 'Khalil']


In [44]:
print(users[-3:len(users)])

['Ahmad', 'Bilal', 'Khalil']


In [45]:
print(users[-4:len(users)])

['Ahmad', 'Bilal', 'Khalil']


In [46]:
print(users[0:0])

[]


In [47]:
print(users[0:-2])

['Ahmad']


In [48]:
print(users[1:-1])

['Bilal']


In [49]:
print(users[2:])

['Khalil']


<b>List <code>Start-Stop-Step</code> Slicing</b>

In [50]:
print(users[0::])

['Ahmad', 'Bilal', 'Khalil']


In [51]:
print(users[1::])

['Bilal', 'Khalil']


In [52]:
print(users[2::])

['Khalil']


In [53]:
print(users[3::])

[]


In [54]:
print(users[0:0:])

[]


In [55]:
print(users[0:1:])

['Ahmad']


In [56]:
print(users[0:2:])

['Ahmad', 'Bilal']


In [57]:
print(users[0:3:])

['Ahmad', 'Bilal', 'Khalil']


In [58]:
print(users[0:4:])

['Ahmad', 'Bilal', 'Khalil']


In [59]:
# print(users[0:0:0])  # ValueError: slice step cannot be zero

In [60]:
print(users[0:3:1])

['Ahmad', 'Bilal', 'Khalil']


In [61]:
print(users[0:3:2])

['Ahmad', 'Khalil']


In [62]:
print(users[0:3:3])

['Ahmad']


In [63]:
print(users[::-1])

['Khalil', 'Bilal', 'Ahmad']


In [64]:
print(users[len(users)-1::-1])

['Khalil', 'Bilal', 'Ahmad']


In [65]:
print(users[1::-1])

['Bilal', 'Ahmad']


In [66]:
print(users[0::-1])

['Ahmad']


<b>Python <code>slice()</code> Built-in Function</b>

Link: https://docs.python.org/3/library/functions.html#slice

Link: https://docs.python.org/3/glossary.html#term-slice

The <code>slice()</code> function returns a <i>slice-object</i> that can use used to slice <i>lists</i>, <i>strings</i>, <i>tuple</i> etc.

The slice object is used to slice a given sequence (list, tuple, string or range) or any object which supports sequence protocol.

The syntax of slice() is:

    slice(start, stop, step)

slice() can take three parameters:

- <b>start (optional)</b> - Starting integer where the slicing of the object starts. Default to <code>None</code> if not provided.
- <b>stop</b> - Integer until which the slicing takes place. The slicing stops at index stop -1 (last element).
- <b>step (optional)</b> - Integer value which determines the increment between each index for slicing. Defaults to None if not provided.

In [67]:
slice(3)

slice(None, 3, None)

In [68]:
# list(slice(3))   # TypeError: 'slice' object is not iterable

In [69]:
print(users[slice(3)])

['Ahmad', 'Bilal', 'Khalil']


In [70]:
print(users[slice(0, 3, 1)])

['Ahmad', 'Bilal', 'Khalil']


In [71]:
print(users[slice(-1, -4, -1)])

['Khalil', 'Bilal', 'Ahmad']


Slicing a List Simple Examples:
- Taking n first elements of a list
- Taking n last elements of a list
- Taking all but n last elements of a list
- Taking every nth-element of a list

### Copying List Elements

- Shallow copy
- Deep copy 

Link: https://docs.python.org/3/library/copy.html#shallow-vs-deep-copy

A <b>shallow copy</b> constructs a new compound object and then (to the extent possible) inserts <i>references</i> into it to the objects found in the original.

A <b>deep copy</b> constructs a new compound object and then, recursively, inserts <i>copies</i> into it of the objects found in the original.

In [72]:
print(users[:])    # This means that the following slice returns a shallow-copy of the list. 
# That means, we can safely modify the new list and it will not affect the initial list.

['Ahmad', 'Bilal', 'Khalil']


In [73]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [74]:
two_users = users[1:]  

In [75]:
two_users[1] = 'Khalil ur Rehman'

In [76]:
print(two_users)

['Bilal', 'Khalil ur Rehman']


In [77]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [78]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [79]:
friends = users[:]  
# It creates a shallow copy of the whole list and is a good shorthand when you need a copy of the original list.

In [80]:
friends[1] = 'M. Bilal'

In [81]:
print(friends)

['Ahmad', 'M. Bilal', 'Khalil']


In [82]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [83]:
print('Id of users:', hex(id(users)))
print('Id of friends:', hex(id(friends)))

Id of users: 0x1b2bef18348
Id of friends: 0x1b2bf110b88


<code>copy()</code> – Returns a shallow copy of the list.

In [84]:
list.copy?

Signature: list.copy(self, /)
Docstring: Return a shallow copy of the list.
Type:      method_descriptor


In [85]:
f1 = users.copy()

In [86]:
f1[0] = 'Ahmad Hassan'

In [87]:
print(f1)

['Ahmad Hassan', 'Bilal', 'Khalil']


In [88]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [89]:
import copy

In [90]:
copy.copy?

Signature: copy.copy(x)
Docstring:
Shallow copy operation on arbitrary Python objects.

See the module's __doc__ string for more info.
File:      c:\programdata\anaconda3\lib\copy.py
Type:      function


In [91]:
f2 = copy.copy(users)

In [92]:
f2[1] = 'H.M.Bilal'

In [93]:
print(f2)

['Ahmad', 'H.M.Bilal', 'Khalil']


In [94]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


<b>Note</b>: In Python, we use <code>=</code> operator to create a copy of an object. You may think that this creates a new object; it doesn't. It only creates a new variable that shares the <i>reference</i> of the original object.

Let's take an example where we create a list named <code>lst1</code> and pass an object reference to <code>lst2</code> using <code>=</code> operator.

In [95]:
lst1 = [4, 6, 8, 12]

In [96]:
lst2 = lst1

In [97]:
lst2[2] = 10

In [98]:
print(lst2)

[4, 6, 10, 12]


In [99]:
print(lst1)

[4, 6, 10, 12]


In [100]:
print('Id of lst1:', hex(id(lst1)))
print('Id of lst1:', hex(id(lst2)))

Id of lst1: 0x1b2bf119048
Id of lst1: 0x1b2bf119048


### Replacing / Modifying / Updating List Elements

Once you've defined a list, you can change individual elements in the list. You do this by referring to the index of the item you want to modify.

In [101]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [102]:
users[2] = 'Khalil-ur-Rehman'

In [103]:
print(users)

['Ahmad', 'Bilal', 'Khalil-ur-Rehman']


In [104]:
print(values)

[5, 2, 4, 8, 7, 4]


In [105]:
values[-1] = 20

In [106]:
print(values)

[5, 2, 4, 8, 7, 20]


In [107]:
values[:3] = [3, 6, 9]  # Replace part of the list

In [108]:
print(values)

[3, 6, 9, 8, 7, 20]


In [109]:
values[:3] = 'xyz'      # Replace part of the list

In [110]:
print(values)

['x', 'y', 'z', 8, 7, 20]


In [111]:
values[:3] = [5]        # Replace and Resize part of the list

In [112]:
print(values)

[5, 8, 7, 20]


In [113]:
values[::2] = [0, 0]  # Replace Every n-th Element

In [114]:
print(values)

[0, 8, 0, 20]


In [115]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4], ['Java', 6]]


In [116]:
prog_langs[3][1], prog_langs[2][1] = prog_langs[2][1], prog_langs[3][1]

In [117]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 5], ['Julia', 2], ['Go', 4], ['Java', 6]]


In [118]:
prog_langs[-1][0] = 'Scala'

In [119]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 5], ['Julia', 2], ['Go', 4], ['Scala', 6]]


In [120]:
prog_langs[-2:-1] = ['Swift', 4]

In [121]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 5], ['Julia', 2], 'Swift', 4, ['Scala', 6]]


In [122]:
users = ['Ahmad', 'Bilal', 'Khalil']

In [123]:
values = [5, 2, 4, 8, 7, 4]

In [124]:
prog_langs = [['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4], ['Java', 6]]

### Inserting Elements from the List

The <code>insert()</code> method inserts an element to the list at a given index.

<code>insert(index, value)</code> – inserts value just before the specified index. Thus after the insertion the new element occupies position index.

The syntax of <code>insert()</code> method is

    list.insert(index, element)

The insert() function takes two parameters:

- <b>index</b> - position where an element needs to be inserted
- <b>element</b> - this is the element to be inserted in the list

In [125]:
list.insert?

Signature: list.insert(self, index, object, /)
Docstring: Insert object before index.
Type:      method_descriptor


In [126]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [127]:
users.insert(2, 'Rizwan')

In [128]:
print(users)

['Ahmad', 'Bilal', 'Rizwan', 'Khalil']


In [129]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4], ['Java', 6]]


In [130]:
prog_langs.insert(3, ['C#', 7])

In [131]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['C#', 7], ['Julia', 5], ['Go', 4], ['Java', 6]]


In [132]:
f3 = ['Anayat', 'Kamal']

In [133]:
users.insert(len(users), f3)

In [134]:
print(users)

['Ahmad', 'Bilal', 'Rizwan', 'Khalil', ['Anayat', 'Kamal']]


In [135]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['C#', 7], ['Julia', 5], ['Go', 4], ['Java', 6]]


In [136]:
prog_langs[2].insert(len(prog_langs[2]), 'Julia')

In [137]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2, 'Julia'], ['C#', 7], ['Julia', 5], ['Go', 4], ['Java', 6]]


### Extending Elements from the List

The <code>extend()</code> extends the list by adding all items of a list (passed as an argument) to the end.

<code>extend(enumerable)</code> – extends the list by appending elements from another enumerable.

The syntax of <code>extend()</code> method is:

    list1.extend(list2)

As mentioned, the <code>extend()</code> method takes a single argument (a list) and adds it to the end.

If you need to add elements of other native datatypes (like tuple and set) to the list, you can simply use:
    
    list.extend(other_type)        or
    list.extend(list(other_type))

In [138]:
list.extend?

Signature: list.extend(self, iterable, /)
Docstring: Extend list by appending elements from the iterable.
Type:      method_descriptor


In [139]:
print(values)

[5, 2, 4, 8, 7, 4]


In [140]:
nums = [12, 4.5, 10]

In [141]:
nums.extend(values)

In [142]:
print(nums)

[12, 4.5, 10, 5, 2, 4, 8, 7, 4]


Lists can also be concatenated with the <code>+</code> operator. Note that this does not modify any of the original lists.

In [143]:
nums1 = nums + [15, 25.5, 9]

In [144]:
print(nums1)

[12, 4.5, 10, 5, 2, 4, 8, 7, 4, 15, 25.5, 9]


In [145]:
t1 = (4, 7)

In [146]:
nums1.extend(t1)

In [147]:
print(nums1)

[12, 4.5, 10, 5, 2, 4, 8, 7, 4, 15, 25.5, 9, 4, 7]


In [148]:
x = [4, 7]

In [149]:
x.extend(range(3))

In [150]:
print(x)

[4, 7, 0, 1, 2]


### Removing Elements from the List

The <code>remove()</code> method removes the first matching element (which is passed as an argument) from the list.

<code>remove(value)</code> – removes the first occurrence of the specified value. If the provided value cannot be found, a
<code>ValueError</code> is raised.

The syntax of the <code>remove()</code> method is:

    list.remove(element)

The <code>remove()</code> method takes a single element as an argument and removes it from the list.

If the <code>element</code> doesn't exist, it throws <i>ValueError: list.remove(x): x not in list</i> exception.

In [151]:
list.remove?

Signature: list.remove(self, value, /)
Docstring:
Remove first occurrence of value.

Raises ValueError if the value is not present.
Type:      method_descriptor


In [152]:
print(users)

['Ahmad', 'Bilal', 'Rizwan', 'Khalil', ['Anayat', 'Kamal']]


In [153]:
users.remove('Rizwan')

In [154]:
print(users)

['Ahmad', 'Bilal', 'Khalil', ['Anayat', 'Kamal']]


In [155]:
users[-1].remove('Anayat')

In [156]:
print(users)

['Ahmad', 'Bilal', 'Khalil', ['Kamal']]


In [157]:
# users.remove('Faizan')   # ValueError: list.remove(x): x not in list

In [158]:
print(values)

[5, 2, 4, 8, 7, 4]


In [159]:
values.remove(4)  # only the first occurrence of element 4 is removed from the list.

In [160]:
print(values)

[5, 2, 8, 7, 4]


<b>Python <code>del</code> Keyword</b>

The syntax of <code>del</code> statement is:

    del obj_name

Here, <code>del</code> is a Python keyword. And, <code>obj_name</code> can be variables, user-defined objects, lists, items within lists, dictionaries etc.

In [161]:
print(nums1)

[12, 4.5, 10, 5, 2, 4, 8, 7, 4, 15, 25.5, 9, 4, 7]


In [162]:
del nums1[-5:]

In [163]:
print(nums1)

[12, 4.5, 10, 5, 2, 4, 8, 7, 4]


In [164]:
print(nums)

[12, 4.5, 10, 5, 2, 4, 8, 7, 4]


In [165]:
del nums[::2]  # remove each n-th element

In [166]:
print(nums)

[4.5, 5, 4, 7]


In [167]:
del nums

In [168]:
del nums1[:]

### Popping Elements from the List

The <code>pop()</code> method removes the item at the given index from the list and returns the removed item.

<code>pop([index])</code> – removes and returns the item at <code>index</code>. With no argument it removes and returns the last element of the list.

The syntax of the <code>pop()</code> method is:

    list.pop(index)

The <code>pop()</code> method takes a single argument (index).

The argument passed to the method is optional. If not passed, the default index <code>-1</code> is passed as an argument (index of the last item).

If the index passed to the method is not in range, it throws <code>IndexError: pop index out of range</code> exception.

In [169]:
list.pop?

Signature: list.pop(self, index=-1, /)
Docstring:
Remove and return item at index (default last).

Raises IndexError if list is empty or index is out of range.
Type:      method_descriptor


In [170]:
print(users)

['Ahmad', 'Bilal', 'Khalil', ['Kamal']]


In [171]:
users.pop(-1)

['Kamal']

In [172]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [173]:
# users.pop(3)   # IndexError: pop index out of range

In [174]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2, 'Julia'], ['C#', 7], ['Julia', 5], ['Go', 4], ['Java', 6]]


In [175]:
prog_langs.pop(3)

['C#', 7]

In [176]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2, 'Julia'], ['Julia', 5], ['Go', 4], ['Java', 6]]


In [177]:
prog_langs[2].pop()

'Julia'

In [178]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4], ['Java', 6]]


In [179]:
prog_langs.pop()

['Java', 6]

In [180]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]]


### Sorting Elements from the List

The <code>sort()</code> method sorts the elements of a given list.

<code>sort()</code> – sorts the list in numerical and lexicographical order and returns <code>None</code>.

The <code>sort()</code> method sorts the elements of a given list in a specific order - Ascending or Descending.

The syntax of <code>sort()</code> method is:
    
    list.sort(key=..., reverse=...)

By default, <code>sort()</code> doesn't require any extra parameters. However, it has two optional parameters:

- <b>reverse</b> - If true, the sorted list is reversed (or sorted in Descending order)
- <b>key</b> - function that serves as a key for the sort comparison

<code>sort()</code> method doesn't return any value. Rather, it changes the original list.

In [181]:
list.sort?

Signature: list.sort(self, /, *, key=None, reverse=False)
Docstring: Stable sort *IN PLACE*.
Type:      method_descriptor


In [182]:
print(values)

[5, 2, 8, 7, 4]


In [183]:
values.sort()

In [184]:
print(values)

[2, 4, 5, 7, 8]


In [185]:
values.sort(reverse=True)

In [186]:
print(values)

[8, 7, 5, 4, 2]


In [187]:
lst = ['Ahmad', 'Bilal', 'Khalil']

In [188]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [189]:
users.sort(reverse=True)

In [190]:
print(users)

['Khalil', 'Bilal', 'Ahmad']


In [191]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]]


In [192]:
prog_langs.sort(key=lambda n: n[1], reverse=True)

In [193]:
print(prog_langs)

[['Julia', 5], ['Go', 4], ['Rust', 3], ['JavaScript', 2], ['Python', 1]]


Alternatively, you can also use Python's in-built function <code>sorted()</code> for the same purpose.

    sorted(list, key=..., reverse=...)

<b>Note</b>: Simplest difference between <code>sort()</code> and <code>sorted()</code> is: <code>sort()</code> doesn't return any value while, <code>sorted()</code> returns an iterable list.

In [194]:
sorted?

Signature: sorted(iterable, /, *, key=None, reverse=False)
Docstring:
Return a new list containing all items from the iterable in ascending order.

A custom key function can be supplied to customize the sort order, and the
reverse flag can be set to request the result in descending order.
Type:      builtin_function_or_method


In [195]:
prog_langs = [['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]] 

In [196]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]]


In [197]:
sorted(prog_langs, key=lambda n: n[1], reverse=False)

[['Python', 1], ['JavaScript', 2], ['Rust', 3], ['Go', 4], ['Julia', 5]]

In [198]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]]


In [199]:
langs = sorted(prog_langs, key=lambda n: n[1], reverse=False)

In [200]:
print(langs)

[['Python', 1], ['JavaScript', 2], ['Rust', 3], ['Go', 4], ['Julia', 5]]


### Reversing List Elements

The <code>reverse()</code> method reverses the elements of a given list.

<code>reverse()</code> – reverses the list in-place and returns <code>None</code>.

The syntax of <code>reverse()</code> method is:

    list.reverse()

In [201]:
list.reverse?

Signature: list.reverse(self, /)
Docstring: Reverse *IN PLACE*.
Type:      method_descriptor


In [202]:
prog_langs = [['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]] 

In [203]:
prog_langs.reverse()

In [204]:
print(prog_langs)

[['Go', 4], ['Julia', 5], ['JavaScript', 2], ['Python', 1], ['Rust', 3]]


In [205]:
print(users)

['Khalil', 'Bilal', 'Ahmad']


In [206]:
users = users[::-1]

In [207]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


### Searching Elements from the List

The <code>index()</code> method searches an element in the list and returns its index.

<code>index(value, [startIndex])</code> – gets the index of the first occurrence of the input value. If the input value is not in the list a <code>ValueError</code> exception is raised. If a second argument is provided, the search is started at that specified index.

The syntax of the <code>index()</code> method is:

    list.index(element)

This method takes a single argument:

- <b>element</b> - element that is to be searched.

In simple terms, the <code>index()</code> method finds the given element in a list and returns its position.

If the same element is present more than once, the method returns the index of the first occurrence of the element.

<b>Note</b>: Index in Python starts from 0, not 1.

In [208]:
users.index?

Signature: users.index(value, start=0, stop=9223372036854775807, /)
Docstring:
Return first index of value.

Raises ValueError if the value is not present.
Type:      builtin_function_or_method


In [209]:
users = ['Ahmad', 'Bilal', 'Khalil']

In [210]:
values = [5, 2, 4, 8, 7]

In [211]:
prog_langs = [['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]] 

In [212]:
print(users.index('Khalil'))

2


In [213]:
# print(users.index('ahmad'))  # ValueError: 'ahmad' is not in list

In [214]:
print(prog_langs.index(['Julia', 5]))

3


In [215]:
print(prog_langs.index(['Julia', 5], 1))

3


In [216]:
# print(prog_langs.index(['Julia', 5], 1, 3))  # ValueError: ['Julia', 5] is not in list

In [217]:
print(prog_langs.index(['Julia', 5], 1, len(prog_langs)))

3


### List Built-in Methods 

In [218]:
# print(help(list))

In [219]:
# list?

List Built-in Methods

    - append(self, object, /)
    - remove(self, value, /)
    - pop(self, index=-1, /)
    - sort(self, /, *, key=None, reverse=False)
    - reverse(self, /)
    - insert(self, index, object, /)
    - extend(self, iterable, /)
    - copy(self, /)
    - index(self, value, start=0, stop=9223372036854775807, /)
    
    - count(self, value, /)
    - clear(self, /)

In [220]:
# print(dir(list))

<b>List Count() Method</b>

The <code>count()</code> method returns the number of occurrences of an element in a list.

<code>count(value)</code> – counts the number of occurrences of some value in the list.

In simple terms, <code>count()</code> method counts how many times an element has occurred in a list and returns it.

The syntax of <code>count()</code> method is:

    list.count(element)

The <code>count()</code> method takes a single argument:

- <b>element</b> - element whose count is to be found.

In [221]:
list.count?

Signature: list.count(self, value, /)
Docstring: Return number of occurrences of value.
Type:      method_descriptor


In [222]:
print(values)

[5, 2, 4, 8, 7]


In [223]:
values.count(4)

1

In [224]:
values.count(10)

0

In [225]:
data = [3, 5, 6, 'Ahmed', 'Bilal', 3, 6, 3, 6]

In [226]:
data[4:].count(3)

2

In [227]:
v1 = [5, 6, 7, [4, 6, 4], 8, 6, 4]

In [228]:
v1.count(4) + v1[3].count(4)

3

In [229]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]]


In [230]:
prog_langs.count(['Julia', 5])

1

<b>List Clear() Method</b>

The <code>clear()</code> method removes all items from the list.

The syntax of <code>clear()</code> method is:

    list.clear()

In [231]:
list.clear?

Signature: list.clear(self, /)
Docstring: Remove all items from list.
Type:      method_descriptor


In [232]:
data.clear()

In [233]:
len(data)

0

In [234]:
del v1[:]

In [235]:
len(v1)

0

### List Built-in Functions 

In [236]:
# print(help(list))

In [237]:
from array import array

In [238]:
arr1 = array('i', [3, 5, 10])

In [239]:
lst1 = list(arr1)

In [240]:
print(lst1)

[3, 5, 10]


In [241]:
list(range(5, 11))

[5, 6, 7, 8, 9, 10]

In [242]:
print(users)

['Ahmad', 'Bilal', 'Khalil']


In [243]:
print(values)

[5, 2, 4, 8, 7]


In [244]:
print(prog_langs)

[['Rust', 3], ['Python', 1], ['JavaScript', 2], ['Julia', 5], ['Go', 4]]


In [245]:
type(users)

list

In [246]:
print(type(users))

<class 'list'>


In [247]:
print(len(values))

5


In [248]:
sorted(values)

[2, 4, 5, 7, 8]

In [249]:
sorted(values, reverse=True)

[8, 7, 5, 4, 2]

In [250]:
list(reversed(users))

['Khalil', 'Bilal', 'Ahmad']

In [251]:
max(values)

8

In [252]:
min(values)

2

In [253]:
sum(values)

26

In [254]:
str(values)

'[5, 2, 4, 8, 7]'

In [255]:
list('python')

['p', 'y', 't', 'h', 'o', 'n']

In [256]:
tuple(values)

(5, 2, 4, 8, 7)

In [257]:
set(values)

{2, 4, 5, 7, 8}

In [258]:
frozenset(values)

frozenset({2, 4, 5, 7, 8})

In [259]:
dict(prog_langs)

{'Rust': 3, 'Python': 1, 'JavaScript': 2, 'Julia': 5, 'Go': 4}

In [260]:
list(enumerate(users))

[(0, 'Ahmad'), (1, 'Bilal'), (2, 'Khalil')]

In [261]:
dict(list(enumerate(users)))

{0: 'Ahmad', 1: 'Bilal', 2: 'Khalil'}

In [262]:
user = iter(users)

In [263]:
next(user)

'Ahmad'

In [264]:
# print(dir(values))

In [265]:
# print(dir(__builtins__))

### Other List Operations

In [266]:
'Bilal' in users

True

In [267]:
'Rizwan' not in users

True

In [268]:
'Bilal' == users[1]

True

In [269]:
users + ['Rizwan']

['Ahmad', 'Bilal', 'Khalil', 'Rizwan']

In [270]:
users * 2  # Replication

['Ahmad', 'Bilal', 'Khalil', 'Ahmad', 'Bilal', 'Khalil']

In [271]:
digits = '1 2 3 4 5 6 7 8 9'

In [272]:
digits.split(' ')

['1', '2', '3', '4', '5', '6', '7', '8', '9']

In [273]:
digits.split(' ') + list('abc')

['1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c']

In [274]:
del digits

## Python Lists - Conditional and Looping Statements

<b>Example 1:</b>

In [275]:
users = ['ahmad', 'bilal', 'khalil']

In [276]:
name = input('Enter name:')

Enter name: Rizwan


In [277]:
if name not in users:
    users.append(name)
    print('add new user ...')
else:
    print('already exist')

add new user ...


In [278]:
for name in reversed(users):
    print(name.title())

Rizwan
Khalil
Bilal
Ahmad


<b>Example 2:</b>

In [279]:
courses = []

In [280]:
course_len = int(input('Enter total number of courses you want to add : '))

Enter total number of courses you want to add :  5


In [281]:
i = 0
while i < course_len:
    course_name = input(f'Enter course name {i+1}:')
    courses.append(course_name)
    i += 1

Enter course name 1: NLP
Enter course name 2: Computer Vision
Enter course name 3: Big Data
Enter course name 4: Machine Learning
Enter course name 5: Cloud


In [282]:
for i in range(len(courses)):
    print(courses[i])

NLP
Computer Vision
Big Data
Machine Learning
Cloud


<b>Example 3:</b>

In [283]:
lang = ['python', 'python', 'c#', 'ruby', 'python', 'java', 'python', 'python', 'rust']

In [284]:
for item in lang[:]:
    if item == 'python':
        lang.remove(item)
else:
    lang.append('python')

In [285]:
print(lang)

['c#', 'ruby', 'java', 'rust', 'python']


In [286]:
i = len(lang)-1
while i >= 0:
    print(lang[i].upper(), end=' ')
    i -= 1

PYTHON RUST JAVA RUBY C# 

## Python List Comprehensions

List comprehensions in Python are concise, syntactic constructs. They can be utilized to generate lists from other lists by applying functions to each element in the list.

List comprehension is an elegant and concise way to create a new list from an existing list in Python.

Link: https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions

A <b>list comprehension</b> creates a new <code>list</code> by applying an expression to each element of an iterable. The most basic form is:

    [ <expression> for <element> in <iterable> ]

There's also an optional <code>if</code> condition:

    [ <expression> for <element> in <iterable> if <condition> ]

    Each <element> in the <iterable> is plugged in to the <expression> if the (optional) <condition> evaluates to true. All results are returned at once in the new list. 
    
Generator expressions are evaluated lazily, but list comprehensions evaluate the entire iterator immediately - consuming memory proportional to the iterator's length.

To create a <code>list</code> of squared integers:

In [287]:
squares = [x * x for x in (1, 2, 3, 4, 5)]

In [288]:
print(squares)

[1, 4, 9, 16, 25]


The <code>for</code> expression sets <code>x</code> to each value in turn from <code>(1, 2, 3, 4)</code>. The result of the expression <code>x * x</code> is appended to an internal <code>list</code>. The internal <code>list</code> is assigned to the variable <code>squares</code> when completed.

#### Happy Learning 😊